bd-crash-handler/src/uploader.rs

```diff
@@ -0,0 +1,282 @@
+#[path = "./uploader_test.rs"]
+mod tests;
+
+use async_trait::async_trait;
+use bd_api::upload::{IntentDecision, TrackedArtifactIntent, TrackedArtifactUpload};
+use bd_api::DataUpload;
+use bd_client_common::file::{read_compressed_protobuf, write_compressed_protobuf};
+use bd_client_common::file_system::FileSystem;
+use bd_proto::protos::client::api::{UploadArtifactIntentRequest, UploadArtifactRequest};
+use bd_proto::protos::client::artifact::artifact_upload_index::Artifact;
+use bd_proto::protos::client::artifact::ArtifactUploadIndex;
+use bd_shutdown::ComponentShutdown;
+use bd_time::{OffsetDateTimeExt, TimeProvider};
+use mockall::automock;
+use std::collections::VecDeque;
+use std::path::PathBuf;
+use std::sync::{Arc, LazyLock};
+use uuid::Uuid;
+
+/// Root directory for all files used for storage and uploading.
+pub static REPORT_DIRECTORY: LazyLock<PathBuf> = LazyLock::new(|| "report_uploads".into());
+
+/// The index file used for tracking all of the individual files.
+pub static REPORT_INDEX_FILE: LazyLock<PathBuf> = LazyLock::new(|| "report_index.pb".into());
+
+type NewUpload = (Uuid, Vec<u8>);
+
+#[automock]
+#[async_trait]
+pub trait Client: Send + Sync {
+  async fn enqueue_upload(&self, contents: Vec<u8>) -> anyhow::Result<Uuid>;
+}
+
+pub struct UploadClient {
+  upload_tx: tokio::sync::mpsc::Sender<NewUpload>,
+}
+
+#[async_trait]
+impl Client for UploadClient {
+  /// Dispatches a payload to be uploaded, returning the associated artifact UUID.
+  async fn enqueue_upload(&self, contents: Vec<u8>) -> anyhow::Result<Uuid> {
+    let report_uuid = uuid::Uuid::new_v4();
+
+    self
+      .upload_tx
+      .try_send((report_uuid, contents))
+      .inspect_err(|e| {
+        log::warn!("failed to enqueue artifact upload: {e}");
+      })?;
+
+    Ok(report_uuid)
+  }
+}
+
+pub struct Uploader {
+  data_upload_tx: tokio::sync::mpsc::Sender<DataUpload>,
+  upload_queued_rx: tokio::sync::mpsc::Receiver<NewUpload>,
+  shutdown: ComponentShutdown,
+  time_provider: Arc<dyn TimeProvider>,
+  file_system: Arc<dyn FileSystem>,
+
+  index: VecDeque<Artifact>,
+}
+
+impl Uploader {
+  pub fn new(
+    file_system: Arc<dyn FileSystem>,
+    data_upload_tx: tokio::sync::mpsc::Sender<DataUpload>,
+    time_provider: Arc<dyn TimeProvider>,
+    shutdown: ComponentShutdown,
+  ) -> (Self, UploadClient) {
+    let (upload_tx, upload_rx) = tokio::sync::mpsc::channel(2);
+
+    let uploader = Self {
+      data_upload_tx,
+      upload_queued_rx: upload_rx,
+      shutdown,
+      time_provider,
+      file_system,
+      index: VecDeque::default(),
+    };
+
+    let client = UploadClient { upload_tx };
+
+    (uploader, client)
+  }
+
+  pub async fn run(mut self) -> anyhow::Result<()> {
+    self.initialize().await?;
+
+    // TODO(snowp): Due to the way this is structured we only pop values off the queue of inbound
+    // requests while we have no pending uploads. It may need to be two tasks that interact with
+    // the index through a mutex.
```

Comment:
FWIW I think this is required. During no network scenarios we will start dropping right away. We should keep accepting uploads up to the limits imposed in the other TODOs.